pkg/openfeaturetest/testprovider.go

```diff
@@ -0,0 +1,104 @@
+// `SetProvider` method.
+type TestAwareProvider struct {
+	openfeature.NoopProvider
+	providers *sync.Map
+}
+
+// SetProvider sets a given `FeatureProvider` for a given test.
+func (tp TestAwareProvider) SetProvider(test *testing.T, fp openfeature.FeatureProvider) {
+	storeGoroutineLocal(testNameKey, test.Name())
+	tp.providers.Store(test.Name(), fp)
+}
+
+func (tp TestAwareProvider) BooleanEvaluation(ctx context.Context, flag string, defaultValue bool, flCtx openfeature.FlattenedContext) openfeature.BoolResolutionDetail {
+	return tp.getProvider().BooleanEvaluation(ctx, flag, defaultValue, flCtx)
+}
+
+func (tp TestAwareProvider) StringEvaluation(ctx context.Context, flag string, defaultValue string, flCtx openfeature.FlattenedContext) openfeature.StringResolutionDetail {
+	return tp.getProvider().StringEvaluation(ctx, flag, defaultValue, flCtx)
+}
+
+func (tp TestAwareProvider) FloatEvaluation(ctx context.Context, flag string, defaultValue float64, flCtx openfeature.FlattenedContext) openfeature.FloatResolutionDetail {
+	return tp.getProvider().FloatEvaluation(ctx, flag, defaultValue, flCtx)
+}
+
+func (tp TestAwareProvider) IntEvaluation(ctx context.Context, flag string, defaultValue int64, flCtx openfeature.FlattenedContext) openfeature.IntResolutionDetail {
+	return tp.getProvider().IntEvaluation(ctx, flag, defaultValue, flCtx)
+}
+
+func (tp TestAwareProvider) ObjectEvaluation(ctx context.Context, flag string, defaultValue interface{}, flCtx openfeature.FlattenedContext) openfeature.InterfaceResolutionDetail {
+	return tp.getProvider().ObjectEvaluation(ctx, flag, defaultValue, flCtx)
+}
+
+func (tp TestAwareProvider) Hooks() []openfeature.Hook {
+	return tp.NoopProvider.Hooks()
+}
+
+func (tp TestAwareProvider) Metadata() openfeature.Metadata {
+	return tp.NoopProvider.Metadata()
+}
+
+func (tp TestAwareProvider) getProvider() openfeature.FeatureProvider {
+	// Retrieve the test name from the goroutine-local storage.
+	testName, ok := getGoroutineLocal(testNameKey).(string)
+	if !ok {
+		panic("unable to detect test name")
+	}
+
+	// Load the feature provider corresponding to the test name.
+	provider, ok := tp.providers.Load(testName)
+	if !ok {
+		panic("unable to find feature provider for given test name: " + testName)
+	}
+
+	// Assert that the loaded provider is of type openfeature.FeatureProvider.
+	featureProvider, ok := provider.(openfeature.FeatureProvider)
+	if !ok {
+		panic("invalid type for feature provider for given test name: " + testName)
+	}
+
+	return featureProvider
+}
+
+var goroutineLocalData sync.Map
+
+func storeGoroutineLocal(key, value interface{}) {
+	gID := getGoroutineID()
+	goroutineLocalData.Store(fmt.Sprintf("%d_%v", gID, key), value)
+}
+
+func getGoroutineLocal(key interface{}) interface{} {
+	gID := getGoroutineID()
+	value, _ := goroutineLocalData.Load(fmt.Sprintf("%d_%v", gID, key))
+	return value
+}
+
+func getGoroutineID() uint64 {
```

Comment:
not perfect but does the trick of getting the current gid from the runtime

Reply:
@warber to be honest it is a very hacky way of dealing with that.

https://go.dev/doc/faq#no_goroutine_id

Reply:
> @warber to be honest it is a very hacky way of dealing with that. Maybe we can think of a more elegant solution [since the go routine id is intentionally not available](https://go.dev/doc/faq#no_goroutine_id).

I agree it's hacky, however, there are multiple relatively popular packages that use this same mechanism, and we are only using it for testing. I would not advocate for including this in production code, but I think it's acceptable to dodge our singleton issue for tests.

IMO this is a net improvement. I'm open to other solutions but at the end of the day, we need some transparent piece of data (at least an id) to propagate from the test to the production code; I'm not aware of any other alternative.

Reply:
ok, I can agree with that